### PR TITLE
KAFKA-6968: Adds calls to listener on rebalance of MockConsumer

### PR DESCRIPTION
**Pull request by efgpinto**
The MockConsumer has a rebalance operation which receives a new set of partitions to simulate the rebalance of a consumer group. However, the current version is not calling the listener provided on subscription to notify the client when partitions are revoked and assigned.

This MR calls both listeners on rebalance and adds a unit test to cover that.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by efgpinto** [comment]
retest this please

**Comment by efgpinto** [comment]
retest this please

**Comment by efgpinto** [comment]
Hey 👋 
If someone can have a look on this. cc @cmccabe @rajinisivaram 
Tks 🙏

**Comment by ocadaruma** [comment]
@efgpinto Hi.
In real KafkaConsumer, rebalance listener is called as part of poll().
To simulate consumer behavior correctly, MockConsumer should call rebalance listener in same place I think. like: https://github.com/efgpinto/kafka/pull/1

What do you think ?

**Comment by efgpinto** [comment]
> @efgpinto Hi.
> In real KafkaConsumer, rebalance listener is called as part of poll().
> To simulate consumer behavior correctly, MockConsumer should call rebalance listener in same place I think. like: [efgpinto#1](https://github.com/efgpinto/kafka/pull/1)
> 
> What do you think ?

Seems like a fair point. Updated. Thanks for your input. 👍

**Comment by baltiyskiy** [comment]
> @efgpinto Hi.
> In real KafkaConsumer, rebalance listener is called as part of poll().
> To simulate consumer behavior correctly, MockConsumer should call rebalance listener in same place I think. like: [efgpinto#1](https://github.com/efgpinto/kafka/pull/1)

What are you talking about? There're no calls to rebalance listeners as far as I can see: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

In the real case, `rebalance()` is called in `ConsumerCoordinator` upon `onJoinComplete()`/`onJoinPrepare()`. `rebalance()` effectively mocks these two. So rebalance listener should be called in `rebalance()`. I don't see any reason to do that in `poll()`.

**Comment by ocadaruma** [comment]
Yeah, rebalance listener is called upon `onJoinComplete/onJoinPrepare`, and it happens as part of `KafkaConsumer#poll` as stated in `ConsumerRebalanceListener`'s javadoc. (https://github.com/apache/kafka/blob/2.5.0/clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerRebalanceListener.java#L48)

as like below:
```
KafkaConsumer#poll() =>
  KafkaConsumer#updateAssignmentMetadataIfNeeded() =>
    ConsumerCoordinator#poll() =>
      AbstractCoordinator#ensureActiveGroup() =>
        AbstractCoordinator#joinGroupIfNeeded() =>
          ConsumerCoordinator#onJoinPrepare()
```

So, if we respect real KafkaConsumer behavior, rebalance listener invocation should be happen in `poll()` rather than `MockConsumer#rebalance()`, which might be called separately from poll().

However, after reconsideration, though my initial intention was like above, I have different thought about this now.

Even if we call rebalance listener in `MockConsumer#rebalance()` (as @efgpinto 's original patch), still we can simulate poll()-driven rebalance invocation (described above) as like below:
```java
@Test
public void testRebalance() {
    MockConsumer<String, String> consumer = new MockConsumer<>();

    consumer.subscribe(Arrays.asList("topic"), new SomeRebalanceListener());

    // simulate rebalance happens after first poll()
    consumer.scheduleNopPollTask();
    consumer.schedulePollTask(() => {
        consumer.rebalance(newAssignment);
    });

    consumer.poll(100L);
    consumer.poll(100L); // => rebalance will be triggered
}
```

So reverting my patch (https://github.com/apache/kafka/pull/7539/commits/4b62aa09eacfea70daa55151ee0928720c772e48) would be fine.

**Comment by baltiyskiy** [comment]
Thanks, you're right about the `poll()`. My concern is that code under test can work in a different thread from the testing code, and it's crucial for `rebalance` to have effect immediately, so that after the call test code can `addRecord()` and expect it to be visible during the next `poll()`. Otherwise, code under test can observe the record before the assignment. For example, https://issues.apache.org/jira/browse/KAFKA-7204 would cause the added record to be dropped, making the test fail. There could be other problems as well.

**Comment by efgpinto** [comment]
Thanks for your insights @ocadaruma and @baltiyskiy . I've reverted back to the previous version.

retest this please

**Comment by efgpinto** [comment]
retest this please

**Comment by CEikermann** [comment]
Would be nice to merge this

Is anything open for this merge request before it can be merged?

**Comment by CEikermann** [comment]
Would be nice to merge this

Is anything open for this merge request before it can be merged?

**Comment by kosiakk** [comment]
Depending on the assignment strategy, not all currently assigned partitions will be rewoked. For example `org.apache.kafka.clients.consumer.CooperativeStickyAssignor` tries to preserve currently assigned partitions.

My Mock subclass replicates that behavour using a simple Set difference calculation. I've just extended the Mock Consumer to overwrite `rebalance` implementation, actually in Kotlin (here just to illustrate the intention)

```Kotlin
    @Synchronized
    override fun rebalance(newAssignment: Collection<TopicPartition>) {
        val listener = lastRebalanceListener ?: error("Please call `subscribe` before `rebalance`")

        val revoked = assignment() - newAssignment
        val assigned = newAssignment - assignment()

        listener.onPartitionsRevoked(revoked)
        super.rebalance(newAssignment)
        listener.onPartitionsAssigned(assigned)
    }
```

or a patch proposal for the main class in Java:
```Java
    public synchronized void rebalance(Collection<TopicPartition> newAssignment) {
        this.records.clear();
        // todo check this.subscriptions.rebalanceListener() for null

        final Set<TopicPartition> revoked = this.subscriptions.assignedPartitions();
        revoked.removeAll(newAssignment);

        final Set<TopicPartition> assigned = new HashSet<>(newAssignment);
        assigned.removeAll(this.subscriptions.assignedPartitions());

        this.subscriptions.rebalanceListener().onPartitionsRevoked(revoked);
        this.subscriptions.assignFromSubscribed(newAssignment);
        this.subscriptions.rebalanceListener().onPartitionsAssigned(assigned);
    }
```

You can actually see this in the log of a real implementation when a second node joins the group:
```
[consumerThread] INFO org.apache.kafka.clients.consumer.internals.ConsumerCoordinator - [Consumer clientId=consumer-test1-1, groupId=test1] Updating assignment with
	Assigned partitions:                       [INPUT-2, INPUT-3]
	Current owned partitions:                  [INPUT-2, INPUT-3, INPUT-0, INPUT-1]
	Added partitions (assigned - owned):       []
	Revoked partitions (owned - assigned):     [INPUT-0, INPUT-1]
```

and then debugger confirms that `onPartitionsRevoked` is called only with **2-element set** with the difference, and then `onPartitionsAssigned` is callsed with an **empty set**.

tl;dr: please don't revoke and then add again unnecessarily, it might be expensive in the app
